org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/telemetry/DocumentTelemetryInfo.java

```diff
@@ -0,0 +1,54 @@
+/*******************************************************************************
+* Copyright (c) 2021 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.telemetry;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.extensions.contentmodel.model.ContentModelManager;
+import org.eclipse.lemminx.extensions.contentmodel.model.ReferencedGrammarInfo;
+
+public class DocumentTelemetryInfo {
+
+	private static final String DOC_PROP_EXT = "file.extension";
+	private static final String DOC_PROP_GRAMMAR = "file.grammar";
+	private static final String DOC_PROP_GRAMMAR_EXTERNAL = "file.grammar.external";
+
+	public static Map<String, Object> getDocumentTelemetryInfo (DOMDocument doc, ContentModelManager manager) {
+		String uri = doc.getDocumentURI();
+		int index = uri.lastIndexOf('.');
+		String fileExtension = uri.substring(index + 1, uri.length());
+		Set<ReferencedGrammarInfo> referencedGrammarInfos = manager.getReferencedGrammarInfos(doc);
+		HashMap<String, Object> props = new HashMap<>();
+		props.put(DOC_PROP_EXT, fileExtension);
+
+		if (referencedGrammarInfos.isEmpty()) {
+			props.put(DOC_PROP_GRAMMAR, "none");
+		} else {
+			List<String> values = new ArrayList<String>();
+			for (ReferencedGrammarInfo info : referencedGrammarInfos) {
+				String kind = info.getIdentifier().getKind();
+				if (kind == null) {
```

Comment:
For some files (eg. lemminx/org.eclipse.lemminx/src/test/resources/xsd/camel-spring-2.22.0.xsd), the kind is null, but the public id can be something like "http://www.w3.org/2001/XMLSchema". Does using this make sense ?

Reply:
Indeed it was not very clear, that's why I created the PR https://github.com/eclipse/lemminx/pull/1098 to simplify the API.

> For some files (eg. lemminx/org.eclipse.lemminx/src/test/resources/xsd/camel-spring-2.22.0.xsd or lemminx/org.eclipse.lemminx/src/test/resources/catalogs/catalog.xml), the kind is null

Yes it's normal, the kind is null when you don't declare the binding in the XML(when you don't use xsi:schemaLocation, xml-model, etc)

> but the public id can be something like "http://www.w3.org/2001/XMLSchema". Does using this make sense ?

This usecase is for XSD. In the case of XSD schema file, the XSD declares the `http://www.w3.org/2001/XMLSchema` namespace (which becomes the public identifier URI while resolving step). The XSD is associated with the xml.xsd file (to manage completion and allowed element,attributes validation for XSD (ex : xs:element is allowed, xs:bad is forbidden). 
The xml.xsd is retrived by the uri resolver extension https://github.com/eclipse/lemminx/blob/cd73f31a4081744b376df65ea5bb2e26464c6ea5/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/XSDURIResolverExtension.java#L52 (the info.getResolvedBy() should return "embedded xml.xsd")

Reply:
Thanks. I've rebased against your new contribution.